src/Hopper/Internal/LoweredCore/ANF.hs

```diff
@@ -0,0 +1,477 @@
+-- 'AnfLet'.
+convertGuarded :: Term -> LoweringM Anf
+convertGuarded t = local (emptyLevel:) $ convertTail t
+
+-- | Converts a 'Term' in tail position to ANF. This function corresponds to
+-- Danvy's function "E".
+convertTail :: Term -> LoweringM Anf
+convertTail term = case term of
+  V v -> do
+    translatedVar <- reader $ translateTermVar v
+    return $ AnfReturn $ V.singleton translatedVar
+
+  BinderLevelShiftUP _ _ ->
+    -- TODO: expose this using proper error machinery
+    error "unexpected binder shift in convertTail during ANF conversion"
+
+  ELit lit ->
+    return $ returnAllocated $ AllocLit lit
+
+  Return terms ->
+    convertToVars (V.toList terms) $ \vars _cleanup ->
+      return $ AnfReturn $ V.fromList vars
+
+  EnterThunk t ->
+    convertToVars [t] $ \[var] _cleanup ->
+      return $ AnfTailCall $ AppThunk var
+
+  Delay t -> do
+    body <- convertGuarded t
+    return $ returnAllocated $ AllocThunk body
+
+  App ft ats -> do
+    let terms = ft : V.toList ats
+    convertToVars terms $ \vars _cleanup ->
+      return $ AnfTailCall $ AppFun (head vars)
+                                    (V.fromList $ tail vars)
+
+  PrimApp primId terms ->
+    convertToVars (V.toList terms) $ \vars _cleanup ->
+      return $ AnfTailCall $ AppPrim primId $ V.fromList vars
+
+  Lam binderInfos t -> do
+    body <- convertGuarded t
+    return $ returnAllocated $ AllocLam (arity binderInfos) body
+
+  Let _binderInfos rhs body ->
+    -- TODO: use binderInfos when we moved to type-directed?
+    convertWithCont rhs TermBinding $ \trackRhs ->
+      local trackRhs $ convertTail body
+
+-- | The total number of binders in the levels of the first stack not shared by
+-- the second stack.
+bindersAddedSince :: BindingStack -> BindingStack -> Word32
+bindersAddedSince extended base = sum $ height <$> extended `levelsSince` base
+  where
+    height :: BindingLevel -> Word32
+    height level | isJust (_levelIndirections level) = _levelIntros level
+                 | otherwise                         = 1 + _levelIntros level
+
+    levelsSince :: BindingStack -> BindingStack -> [BindingLevel]
+    levelsSince new old | new == old = []
+    levelsSince [] _old = error "first stack must be an extension of the second"
+    levelsSince (level : newRest) old = level : (newRest `levelsSince` old)
+
+-- | The former with open binders increased for each extra binder present
+-- in the latter. Assumes that the latter is the former extended with extra top
+-- 'BindingLevel's.
+withBinderIncreasesPer :: BindingStack -> BindingStack -> BindingStack
+withBinderIncreasesPer base extended =
+  base & _head.levelIntros                            +~ numBinders
+       & _head.levelRefs.mapped.localNameless.lnDepth +~ numBinders
+  where
+    numBinders = extended `bindersAddedSince` base
+
+-- | Converts a 'Term' in nontail position to ANF. This function corresponds to
+-- Danvy's function "E_c".
+convertWithCont :: Term
```

Comment:
might be good to have both danvy functions grouped together, plus a little stanza right above them to contextualize